flowmachine/flowmachine/features/subscriber/contact_balance.py

```diff
@@ -69,59 +67,67 @@ def __init__(
         subscriber_subset=None,
     ):
-        """
-        """
-        logger.warning(
-            "The the ContactBalance() feature uses CDRs "
-            + "IDs for calculating an subscriber's graph. "
-            + "If IDs are not generated correctly, this "
-            + "feature could yield erroneous results."
-        )
-
-        self.table = table
+        self.tables = tables
         self.start = start
         self.stop = stop
         self.hours = hours
-        self.subscriber_identifier = subscriber_identifier
         self.direction = direction
+        self.subscriber_identifier = subscriber_identifier
         self.exclude_self_calls = exclude_self_calls
 
-        if self.direction not in ("both", "in", "out"):
-            raise ValueError("Unidentified direction: {}".format(self.direction))
+        if self.direction == "both":
+            column_list = [self.subscriber_identifier, "msisdn_counterpart"]
+            self.tables = tables
+        else:
+            column_list = [self.subscriber_identifier, "msisdn_counterpart", "outgoing"]
+            self.tables = self._parse_tables_ensuring_direction_present(tables)
 
-        cols = [self.subscriber_identifier, "id", "msisdn_counterpart", "outgoing"]
         self.unioned_query = EventsTablesUnion(
             self.start,
             self.stop,
-            columns=cols,
-            tables=self.table,
+            columns=column_list,
+            tables=self.tables,
             subscriber_identifier=self.subscriber_identifier,
             hours=hours,
             subscriber_subset=subscriber_subset,
         ).get_query()
         self._cols = ["subscriber", "msisdn_counterpart", "events", "proportion"]
         super().__init__()
 
+    def _parse_tables_ensuring_direction_present(self, tables):
+
+        if isinstance(tables, str) and tables.lower() == "all":
+            tables = [f"events.{t}" for t in self.connection.subscriber_tables]
+        elif type(tables) is str:
+            tables = [tables]
+        else:
+            tables = tables
+
+        parsed_tables = []
+        tables_lacking_direction_column = []
+        for t in tables:
+            if "outgoing" in Table(t).column_names:
+                parsed_tables.append(t)
+            else:
+                tables_lacking_direction_column.append(t)
+
+        if tables_lacking_direction_column:
+            raise MissingDirectionColumnError(tables_lacking_direction_column)
+
+        return parsed_tables
+
     def _make_query(self):
 
         filters = []
-        if self.direction == "in":
+        if self.direction != "both":
             filters.append(
-                """
-            AND outgoing = FALSE
-            """
-            )
-        elif self.direction == "out":
-            filters.append(
-                """
-            AND outgoing = TRUE
-            """
+                f"outgoing IS {'TRUE' if self.direction == 'out' else 'FALSE'}"
```

Comment:
See above, I'm not super keen on unconditional `else` clauses that act as a catch-all when we are expecting a particular value (here: `self.direction == 'in'`). If the code in `__init__` already did the check that `direction` was only one of the three expected values than this would be kind-of ok but here it could be an arbitrary value and would land in the `else` branch.

Really what this is hinting at is again that `direction` should perhaps not be a string but a slightly smarter object (or an enum). But I'd leave that for another PR. For the time being it should be ok to just check valid values in `__init__` (see comment above).